vmm-reference-interrupts.md

```diff
@@ -0,0 +1,113 @@
+
+There are two helper functions defined to read and write to the LAPIC registers. The `read_le_i32` function reads from the LAPIC register at the given offset and returns the value read. 
+```rust
+fn read_le_i32(input: &[i8]) -> i32 {
+    assert_eq!(input.len(), 4);
+    let mut array = [0u8; 4];
+    for (byte, read) in array.iter_mut().zip(input.iter().cloned()) {
+        *byte = read as u8;
+    }
+    i32::from_le_bytes(array)
+}
+```
+
+The `write_le_i32` function writes the given value to the LAPIC register at the given offset.
+```rust
+fn write_le_i32(buf: &mut [i8], n: i32) {
+    for (byte, read) in buf.iter_mut().zip(i32::to_le_bytes(n).iter().cloned()) {
+        *byte = read as i8;
+    }
+}
+```
+The helper functions `get_klapic_reg` uses the `read_le_i32` function to read from the LAPIC register at the given offset and returns the value read. The values are stored as `i32` which requires 4 `i8` values. Hence, the range is defined as `reg_offset..reg_offset + 4`. 
+
+```rust
+pub fn get_klapic_reg(klapic: &kvm_lapic_state, reg_offset: usize) -> Result<i32> {
+    let range = reg_offset..reg_offset + 4;
+    let reg = klapic.regs.get(range).ok_or(Error::InvalidRegisterOffset)?;
+    Ok(read_le_i32(reg))
+}
+```
+
+Similary `set_klapic_reg` uses the `write_le_i32` function to write the given value to the LAPIC register at the range offset, offset+4. 
+```rust
+pub fn set_klapic_reg(klapic: &mut kvm_lapic_state, reg_offset: usize, value: i32) -> Result<()> {
+    // The value that we are setting is a u32, which needs 4 bytes of space.
+    // We're here creating a range that can fit the whole value.
+    let range = reg_offset..reg_offset + 4;
+    let reg = klapic
+        .regs
+        .get_mut(range)
+        .ok_or(Error::InvalidRegisterOffset)?;
+    write_le_i32(reg, value);
+    Ok(())
+}
+```
+## Error Handling
+There is only one error defined as `InvalidRegisterOffset` and is returned when the register offset is invalid. The check is done after accessing registers and error is returned in case of invalid register offset. 
+```rust
+#[derive(Debug, PartialEq, thiserror::Error)]
+pub enum Error {
+    /// The register offset is invalid.
+    #[error("The register offset is invalid.")]
+    InvalidRegisterOffset,
+}
+```
+
+## Usage in code
+
+While creating a new ```kvmVcpu```, LAPICs for the vcpu is created and initialized with ```configure_lapic``` function. The function ```configure_lapic``` initializes the LAPICs with the following values: 
+1. `LAPIC0`: external interrupts 
+2. `LAPIC1`: NMI.
+
+```rust
+    fn configure_lapic(&self) -> Result<()> {
+        let mut klapic: kvm_lapic_state = self.vcpu_fd.get_lapic().map_err(Error::KvmIoctl)?;
+        set_klapic_delivery_mode(&mut klapic, APIC_LVT0_REG_OFFSET, DeliveryMode::ExtINT).unwrap();
+        set_klapic_delivery_mode(&mut klapic, APIC_LVT1_REG_OFFSET, DeliveryMode::NMI).unwrap();
+        self.vcpu_fd.set_lapic(&klapic).map_err(Error::KvmIoctl)
+    }
+```
```

Comment:
This could be moved to the markdown file that explains the parent rust file that holds `configure_lapic`. `configure_lapic` explanation in the other file can keep a pointer to this file to know more about LAPICs.